### PR TITLE
smart-queue next iteration WIP

### PR DESCRIPTION
**Pull request by calumpeak**
WORK IN PROGRESS

Wanted to ping this up to get some feedback on the direction that this is going (adapting @prsn implementation) and continue to work on with feedback. I've not completely tested this yet so YMMV. 

Addressing https://github.com/redux-offline/queue/issues/4 And #4 

Changes:
 - Correctly add new entities to queue in scenarios where a merge isn't possible
 - Fix immutability issues
 - Allow merge strategy to be provided in config
 - Remove error throwing and use console.error/warn to avoid breaking middleware
 - Update merging strategies for CRUD updates as per addressed issues

TODO:
 - ~~Tests~~
 - ~~Types/Flow~~
 - Merging strategy for base use cases needs full implementation to support optional fields (thinking I might use `lodash.get` to simplify this) and body/json types on effect

### CODE REVIEW

File: packages/smart-queue/src/utils.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * Checks whether an action is safe to be operated on in the queue
+ * if the action is the first in the queue and redux-offline is currently
+ * busy, assume that the action is currently being resolved and is not safe
+ * to operate on
+ * @param {number} index
+ * @param {object} context
+ * @return {boolean}
+ */
+export const safeToProceed = (index, { offline }) =>
+  index !== 0 && !offline.busy;
```
Comment (prsn) [inline packages/smart-queue/src/utils.js]: @echoes221 ,
I think function `safeToProceed` will return `false` for every case other than `non-zero index` and when `offline is not busy`, but if `offline is not busy` then we can consider `zero` index to be a safe index to merge. Moreover, any `non-zero` index can be considered to be safe irrespective of `offline.busy` status.

@sorodrigo, @echoes221  thoughts?

**Review by prsn — COMMENTED** [review]
Hi @echoes221 ,
The way you have structured the code base it's an amazing work man.

File: packages/smart-queue/src/utils.js (hunk 1 above)
Comment (calumpeak) [reply packages/smart-queue/src/utils.js]: Ah! That should be index===0. Thanks for the catch, I'm still writing tests, that should cover that implementation.

Comment (calumpeak) [reply packages/smart-queue/src/utils.js]: So I think this needs to be `!(index === 0 && offline.busy)`. Need to flip it to make it 'safe'. Should cover off the use cases above.

**Comment by calumpeak** [comment]
@prsn thanks man :) Glad you like 👍

**Comment by sorodrigo** [comment]
Thanks for the great work I haven't had the chance to take a look into this carefully. I will try tomorrow!

**Comment by calumpeak** [comment]
Updated majority of tests, only one missing now is for the merge strategy.

File: packages/smart-queue/__tests__/enqueue.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  ],
+  [
+    'it merges UPDATE actions into existing CREATE action with the same key into outbox',
+    createAction(UPDATE, 1, { bar: 2 }),
+    [read1, createAction(CREATE, 1, { foo: 1, bar: 2 }), update1, delete1]
+  ],
+  [
+    'it removes CREATE outbox action when incoming DELETE action has the same key from outbox',
+    createAction(DELETE, 1),
+    [read1, update1, delete1]
```
Comment (sorodrigo) [inline packages/smart-queue/__tests__/enqueue.js:69]: I think the delete1 shouldn't be in the outbox. If we remove the create1 there's no point on performing a delete1 on our backend

**Comment by sorodrigo** [comment]
Ok looks amazing great job!! I wonder have you thought about what to do with the merge function `TODOs`? Maybe I can help you test the remaining fn, I just need to know how to handle the `TODOs`.

File: packages/smart-queue/__tests__/enqueue.js (hunk 2 above)
Comment (calumpeak) [reply packages/smart-queue/__tests__/enqueue.js:69]: It's a little confusing, delete1 is Id 4 as per the constant setup, the delete coming in is ID 1. It's making sure that we don't override another delete. Maybe it's worth updating the constants to reflect their ID?

**Comment by calumpeak** [comment]
@sorodrigo I'm glad you like!
 I've got most of the types done, I'm just having some difficulty with with the callableObject type and how to reference that correctly - I might have to make the QueueFuncs IIFE to type them correctly. 

The type: 
```javascript
export type QueueFunction = {
  (Outbox, OfflineAction, number, Config): Outbox,
  allowedMethods: Method[]
};
```
Using that is a bit awkward on a queue function:
```javascript
// This errors
export const readInQueue: QueueFunction = (outbox, action, index, { merge }) =>
  replaceAtIndex(outbox, merge(outbox[index], action), index);
readInQueue.allowedMethods = [READ];
```
Errors because `allowedMethods` is not defined at that point. So I can either extend the function `(Object.assign(func, { allowedMethods })`, or wrap them both in an IIFE. Unless you've got a better insight into this?

I've had some difficulty with flow and my IDE, it's a little clunky and slow compared to my (limited) experience with TS. I would suggest to move to TS to prevent us needing to manually type the `t.ds` file also, but redux-offline core is in flow so I'll save that for another conversation/next version (especially since babel7 has TS support baked in) :D 

Regarding the merge strategy any help would be appreciated. I've left a couple of TODO's on the function itself. It's more about handling the optional `meta.offline` keys, as well as providing support for `effect.body` and `effect.json`
I thought about using lodash.get as that would allow us to statically type paths, and provide defaults on the fly without tonnes of conditionals.

**Comment by calumpeak** [comment]
So I settled on the Object.assign method for typing the callable objects. Not as neat, but it works.

Just the merging strategy to solidify. I'll have a think on it to see if I can come up with a smart way of doing it.

File: packages/smart-queue/src/enqueue.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    if (!valid(queueMeta)) {
+      // eslint-disable-next-line no-console
+      console.error(new Error(WARNING));
+      return outbox;
+    }
+
+    const queueFunction = getQueueFunction(queueMeta);
+    const index = indexOfAction(outbox, queueMeta);
```
Comment (calumpeak) [inline packages/smart-queue/src/enqueue.js:33]: `indexOfAction` should also contain the check for `methodAllowed` as certain actions cannot be squashed over each other. We should aim to find the best available, not give up on the first id match which may not allow a merge strategy to take place. E.g. `enqueue([Read, Update], Update) -> [Read, Update, Update]`, first check fails so we add to the queue as new entity, where what we actually want is `[Read, Update]`

**Pull request closed without merging**